--- a/plugin/smartdns-ui/Makefile
+++ b/plugin/smartdns-ui/Makefile
@@ -27,7 +27,7 @@ ifeq ($(origin CC), environment)
     ifeq ($(CARGO_BUILD_ARGS),)
       # find default target by compiler
       ARCH_TARGET:=$(shell $(CC) -dumpmachine)
-      override CARGO_BUILD_TARGET:=$(shell rustc --print target-list | grep $(ARCH_TARGET) | head -n 1)
+      override CARGO_BUILD_TARGET:=$(shell rustc --print target-list | grep -m1 -F $(ARCH_TARGET))
 
       ifeq ($(CARGO_BUILD_TARGET),)
         # not found, try to find target by compiler
@@ -36,14 +36,14 @@ ifeq ($(origin CC), environment)
         ABI=$(shell echo $(ARCH_TARGET) | cut -d - -f 3)
         ifneq ($(ABI),)
           # force set target to $(ARCH)-unknown-linux-$(ABI)
-          override CARGO_BUILD_TARGET:=$(shell rustc --print target-list | grep $(ARCH) | grep linux | grep $(ABI) | grep unknown | head -n 1)
+          override CARGO_BUILD_TARGET:=$(shell rustc --print target-list | grep -F $(ARCH) | grep linux | grep -F $(ABI) | grep unknown | head -n 1)
         endif
       endif
 
       ifneq ($(CARGO_BUILD_TARGET),)
         ARCH_TARGET_PATH=$(CARGO_BUILD_TARGET)/
         override CARGO_RUSTFLAGS=-C linker=$(CC)
-         CARGO_BUILD_ARGS +=--target=$(CARGO_BUILD_TARGET)
+        CARGO_BUILD_ARGS += --target=$(CARGO_BUILD_TARGET)
       endif
     endif
   endif
@@ -57,7 +57,7 @@ ifeq ($(origin CC), environment)
       # find sysroot, add sysroot to BINDGEN_EXTRA_CLANG_ARGS
       SYSROOT:=$(SYSROOT)/..
       override CARGO_BUILD_ENV += BINDGEN_EXTRA_CLANG_ARGS="--sysroot=$(SYSROOT)"
-      CARGO_BUILD_ARGS +=--target=$(CARGO_BUILD_TARGET)
+      CARGO_BUILD_ARGS += --target=$(CARGO_BUILD_TARGET)
     endif
   endif
 endif
@@ -67,22 +67,22 @@ ifneq ($(CARGO_BUILD_TARGET),)
 endif
 
 # check and install bindgen
-IS_BINDGEN_INSTALL=$(shell which bindgen 2>/dev/null)
-ifeq ($(IS_BINDGEN_INSTALL),)
+IS_BINDGEN_INSTALLED=$(shell command -v bindgen >/dev/null 2>&1 && echo 1 || echo 0)
+ifeq ($(IS_BINDGEN_INSTALLED),0)
   $(shell cargo install --force --locked bindgen-cli)
 endif
 
 ifneq ($(CARGO_BUILD_TARGET),)
-  CARGO_BUILD_TARGET_INSTALL=$(shell rustup target list | grep $(CARGO_BUILD_TARGET) | grep installed)
+  CARGO_BUILD_TARGET_INSTALL=$(shell rustup target list | grep -F "$(CARGO_BUILD_TARGET) (installed)")
   ifeq ($(CARGO_BUILD_TARGET_INSTALL),)
     # install target
     $(shell rustup target add $(CARGO_BUILD_TARGET))
   endif
 endif
 
-IS_MUSL=$(shell $(CC) -v 2>&1 | grep -i musl >/dev/null 2>&1 && echo 1 || echo 0)
-ifeq ($(IS_MUSL), 1)
-override CARGO_RUSTFLAGS +=-C target-feature=-crt-static
+IS_MUSL=$(shell $(CC) -v 2>&1 | grep -iq musl && echo 1 || echo 0)
+ifeq ($(IS_MUSL),1)
+  override CARGO_RUSTFLAGS += -C target-feature=-crt-static
 endif
 
 ifdef DEBUG
@@ -91,10 +91,10 @@ ifdef DEBUG
 else
   ifdef OPTIMIZE_SIZE
     CARGO_BUILD_ARGS += --profile release-optmize-size
-	CARGO_BUILD_PATH=target/$(ARCH_TARGET_PATH)release-optmize-size
+    CARGO_BUILD_PATH=target/$(ARCH_TARGET_PATH)release-optmize-size
   else
-    CARGO_BUILD_ARGS +=--release
-  	CARGO_BUILD_PATH=target/$(ARCH_TARGET_PATH)release
+    CARGO_BUILD_ARGS += --release
+    CARGO_BUILD_PATH=target/$(ARCH_TARGET_PATH)release
   endif
 
   SMARTDNS_BUILD_TYPE=
@@ -109,7 +109,7 @@ test-prepare:
 
 $(BIN):
 	MAKEFLAGS= RUSTFLAGS="$(CARGO_RUSTFLAGS) $(RUSTFLAGS)" $(CARGO_BUILD_ENV) cargo build $(CARGO_BUILD_ARGS) --features "build-release"
-	cp $(CARGO_BUILD_PATH)/libsmartdns_ui.so target/ 
+	cp $(CARGO_BUILD_PATH)/libsmartdns_ui.so target/
 
 install: $(BIN)
 	install -v -m 0644 -D -t $(DESTDIR)$(SLIBDIR)/smartdns target/libsmartdns_ui.so
