--- a/plugin/smartdns-ui/Makefile
+++ b/plugin/smartdns-ui/Makefile
@@ -88,7 +88,7 @@ test-prepare:
 
 $(BIN):
 	MAKEFLAGS= RUSTFLAGS="$(CARGO_RUSTFLAGS) $(RUSTFLAGS)" $(CARGO_BUILD_ENV) cargo build $(CARGO_BUILD_ARGS) --features "build-release"
-	cp $(CARGO_BUILD_PATH)/libsmartdns_ui.so target/ 
+	find . -type f -path "*/release/libsmartdns_ui.so" -exec cp {} target/ \; 
 
 install: $(BIN)
 	install -v -m 0644 -D -t $(DESTDIR)$(SLIBDIR)/smartdns target/libsmartdns_ui.so
